### PR TITLE
api: storage_service/natural_endpoints: add tablets support

### PR DESCRIPTION
**Pull request by denesb**
This API endpoint currently returns with status 500 if attempted to be called for a table which uses tablets. This series adds tablet support. No change in usage semantics is required, the endpoint already has a table parameter.
This endpoint is the backend of `nodetool getendpoints` which should now work, after this PR.

Fixes: #17313

### CODE REVIEW

File: replica/database.hh [hunk 1]
```diff
@@ -1316,6 +1316,10 @@ public:
         return _replication_strategy;
     }
 
+    bool uses_tablets() const {
+        return _replication_strategy->uses_tablets();
+    }
+
```
Comment (avikivity) [inline replica/database.hh:1322]: We'll have to rethink it when we go vnode->tablet migration, but that applies to many things.

File: service/storage_service.cc [hunk 2]
```diff
@@ -6246,7 +6246,15 @@ storage_service::get_natural_endpoints(const sstring& keyspace,
     auto schema = _db.local().find_schema(keyspace, cf);
     partition_key pk = partition_key::from_nodetool_style_string(schema, key);
     dht::token token = schema->get_partitioner().get_token(*schema, pk.view());
-    return _db.local().find_keyspace(keyspace).get_effective_replication_map()->get_natural_endpoints(token);
+    const auto& ks = _db.local().find_keyspace(keyspace);
+    if (ks.uses_tablets()) {
+        const auto& topology = get_token_metadata().get_topology();
+        const auto& tm = get_token_metadata().tablets().get_tablet_map(schema->id());
```
[the remark this answers is not in the document]
Comment (denesb) [reply service/storage_service.cc:6252]: Token metadata contains the metadata for *all* keyspaces (and now tables).

Technically what you suggested would work. I could try to just look up `schema->id()` in the `tablet_metadata`, but this API is not so friendly if the table doesn't have tablets. `get_tablet_map()` will just throw.

**Comment by denesb** [comment]
Added `@pytest.mark.xfail()` to the new tests. The `rest_api` test suite doesn't have tablets enabled, and without tablets the test fails. Enabling tablets fails many tests in the suite, see https://github.com/scylladb/scylladb/issues/17338.
I will remove the `xfail` once the test suite is fixed to work with tablets.

File: service/storage_service.cc [hunk 3]
```diff
@@ -6250,7 +6250,15 @@ storage_service::get_natural_endpoints(const sstring& keyspace,
...
     partition_key pk = partition_key::from_nodetool_style_string(schema, key);
     dht::token token = schema->get_partitioner().get_token(*schema, pk.view());
-    return _db.local().find_keyspace(keyspace).get_vnode_effective_replication_map()->get_natural_endpoints(token);
+    const auto& ks = _db.local().find_keyspace(keyspace);
+    if (ks.uses_tablets()) {
+        const auto& topology = get_token_metadata().get_topology();
+        const auto& tm = get_token_metadata().tablets().get_tablet_map(schema->id());
+        return boost::copy_range<inet_address_vector_replica_set>(tm.get_tablet_info(token).replicas | boost::adaptors::transformed([&topology] (const locator::tablet_replica& r) {
+            return topology.get_node(r.host).endpoint();
+        }));
```
Comment (tgrabiec) [inline service/storage_service.cc:6259]: Better to use get_natural_endpoints() from table's erm. For symmetry with vnode case, and brevity.

**Review by tgrabiec — CHANGES_REQUESTED** [review]

**Comment by denesb** [comment]
new in v2:
* test is made xfail (see https://github.com/scylladb/scylladb/pull/17316#issuecomment-1943640141)
* call get_natural_endpoints() on the tables erm, instead of hand-coding it

File: service/storage_service.cc (hunk 3 above)
Comment (denesb) [reply service/storage_service.cc]: Fixed in v2

**Pull request merged**